src/EditorFeatures/Test/SolutionCrawler/WorkCoordinatorTests.cs

```diff
@@ -871,9 +874,31 @@ End Property
 
                 await WaitWaiterAsync(workspace.ExportProvider);
 
-                // mutate solution
-                workspace.OnSolutionAdded(solution);
+                // block solution cralwer from processing.
+                var globalOperation = workspace.Services.GetService<IGlobalOperationNotificationService>();
+                using (var operation = globalOperation.Start("Block SolutionCrawler"))
+                {
+                    // mutate solution
+                    workspace.OnSolutionAdded(solution);
+
+                    // wait for workspace events to be all processed
+                    var workspaceWaiter = GetListenerProvider(workspace.ExportProvider).GetWaiter(FeatureAttribute.Workspace);
+                    await workspaceWaiter.CreateWaitTask();
+
+                    // now wait for semantic processor to finish
+                    var crawlerListener = (AsynchronousOperationListener)GetListenerProvider(workspace.ExportProvider).GetListener(FeatureAttribute.SolutionCrawler);
+
+                    // first, wait for first work to be queued.
+                    await crawlerListener.CreateWaitTask(tokens => tokens.Any(token => token.Name == "EnqueueWorkItemAsync"));
```

Comment:
:bulb: The semantics of this `CreateWaitTask` overload are not obvious when reading the code. A better approach (which covers what we need here but may be expanded further in the future) would be the following:

```csharp
/// <summary>
/// Asynchronously waits for all operations matching the specified predicate to complete.
/// </summary>
public async Task CreateWaitTask(Predicate<DiagnosticAsyncToken> selector)
```

💭 Testing for `token.Name` seems likely to break in the future. It is generally understood that changing the name of a listener is a sensitive operation, but changing one of the token properties is not going to cause problems. This is especially problematic because tests may or may not reveal the problem if the name is changed. If you are going to keep this approach, consider defining a static class `PredefinedAsynchronousOperationNames` which has a documented field `EnqueueWorkItemAsync`, and reference this field in all locations where it has special meaning in tests.

Reply:
I am not sure changing Func<...,bool> to predicate is any better. not sure we ever used predicate instead of func<..,bool>. but selector seems better name and doc comment.

about Token.Name. I think it is fine as it is. test already uses internal implementation to test the code. and creating predefined... just for this test is overkill.

Reply:
not sure I am following. so you want me to change name to "WaitUntilConditionIsMet"?

Reply:
No, I'm looking to make the code look like this instead:

```csharp
// first, wait for first work to be queued.
await crawlerListener.CreateWaitTask(token => token.Name == "EnqueueWorkItemAsync");

// and then wait them to be processed
await crawlerListener.CreateWaitTask(token => token.Tag == workspace);
```

Reply:
I can add comment. but I like the one I have better. more flexible. and I dont need one that you suggested above.